File: scripts/insert_card.js
class MagicDB{
    async invoke() {
        const settings = {
            language: "Italian"
        };

        const url = "https://api.magicthegathering.io/v1/cards";
    
        const cardName = await this.get_input();
    
        new Notice(`Searching for card: ${cardName}`);

        const params = {
            name: cardName,
            language: settings.language
        };
    
        try {
            // console.log(encodeURI(url+"?name="+cardName+"&language="+settings.language));

            // const response = await fetch(encodeURI(url+"?name="+cardName+"&language="+settings.language));
            const response = await fetch(url+"?"+new URLSearchParams({
                name: cardName,
                language: settings.language
            }));
    
            // Do something with the data
            const jsonData = await response.json();

            new Notice("Got a response.");

            const fileContent = JSON.stringify(jsonData);
            await app.vault.adapter.write("latest.json", fileContent);

            const outputDict = {};

            if (jsonData.cards.length === 0) {
                new Notice(`No results :(`);
                return;
            }

            let accepted=false;
            let number_of_cards=jsonData.cards.length;
            let i=1;

            for (const card of jsonData.cards) {
                if (settings.language !== "English" && !card.foreignNames) {
                    continue;
                }
                const localizedData = card.foreignNames.find(value => value.language === settings.language);
                if (!localizedData) {
                    localizedData = card;
                }

                for (const dict of [card, localizedData]) {
                    for (const key in dict) {
                        if (dict.hasOwnProperty(key) && key!="foreignNames" && dict[key] !== null) {
                            if(key=="text" || key=="flavor" || key=="originalText"){
                                outputDict[key]=dict[key].replace(/"/g, "'").replace(/\n/g, " // ");
                            }else if(key=="imageUrl"){
                                outputDict["image_url"]=dict["imageUrl"]
                            }else if(key=="manaCost"){
                                outputDict["mana_cost"]=dict["manaCost"]
                            }else if(key!="legalities"){
                                outputDict[key] = dict[key];
                            }
                        }
                    }
                }

                accepted = await this.get_bool_input("Accept? ("+i+"/"+number_of_cards+")",localizedData.imageUrl);
                
                if (accepted) {
                    break;
                }
                i++;
            }

            if (!accepted) {
                return
            }

            if (outputDict.mana_cost) {
                const manaCost = outputDict.mana_cost;

                if (manaCost[1] !== "W" && manaCost[1] !== "B" && manaCost[1] !== "R" && manaCost[1] !== "G" && manaCost[1] !== "U") { // The first value is a number
                    outputDict.neutral = parseInt(manaCost.split("{")[1], 10);
                }

                outputDict.white = manaCost.split("{W}").length - 1;
                outputDict.black = manaCost.split("{B}").length - 1;
                outputDict.red = manaCost.split("{R}").length - 1;
                outputDict.green = manaCost.split("{G}").length - 1;
                outputDict.blue = manaCost.split("{U}").length - 1;
            }

            let outputText = "---\n";

            for (const key in outputDict) {
                if (typeof outputDict[key] === "string") {
                    outputText += `${key}: "${outputDict[key]}"\n`;
                }else if(Array.isArray(outputDict[key])){
                    outputText += `${key}: ${JSON.stringify(outputDict[key])}\n`;
                }
                else {
                    outputText += `${key}: ${outputDict[key]}\n`;
                }
            }

            outputText += "---";

            new Notice("Writing to file...");

            const filePath = `Cards/${outputDict.name}.md`;
            const file = app.vault.getAbstractFileByPath(filePath);
            if (file) {
                await app.vault.adapter.write(filePath, outputText);
            } else {
                await app.vault.create(filePath, outputText);
            }

            new Notice("Success!");
        } catch (error) {
            console.error(error);
        }
    }

    async get_input(message = "Input Card Name",extra_text="") {
        const {Modal, Setting} = customJS.obsidian;
        return new Promise((resolve, reject) => {
            class MyModal extends Modal {
                constructor() {
                    super(app);
                    this.value = "";
                }

                onOpen() {
                    const {contentEl} = this;
                    contentEl.createEl("h1", {text: message});

                    if (extra_text) {
                        const p = contentEl.createEl("p");
                        p.innerText = extra_text;
                    }
                    

                    const inputSetting = new Setting(contentEl);
                    inputSetting.addText((text) => {
                        text.onChange((value) => {
                            this.value = value;
                        });
                        text.inputEl.focus();
                    });
                    inputSetting.addButton((btn) => btn
                        .setButtonText("Submit")
                        .setCta()
                        .onClick(() => {
                            resolve(this.value);
                            this.close();
                        })
                    );
                    contentEl.addEventListener("keydown", (e) => {
                        if (e.key === "Enter") {
                            e.preventDefault();
                            resolve(this.value);
                            this.close();
                        }
                    });
                }
            }

            new MyModal(app).open();
        });
    }

    async get_bool_input(message = "Input Card Name",extra_text=""){
        const {Modal, Setting} = customJS.obsidian;
        return new Promise((resolve, reject) => {
            class MyBoolInputModal extends Modal {
                constructor(app) {
                    super(app);
                }

                onOpen() {
                    const {contentEl} = this;
                    contentEl.createEl("h1", {text: message});

                    if (extra_text) {
                        const img = contentEl.createEl("img");
                        img.src = extra_text;
                        img.style.width = "250px";
                        img.style.display = "block";
                        img.style.margin = "auto";
                    }

                    const setting = new Setting(contentEl);
                    setting.addButton((btn) => btn
                        .setButtonText("Yes")
                        .setCta()
                        .onClick(() => {
                            resolve(true);
                            this.close();
                        })
                    );
                    setting.addButton((btn) => btn
                        .setButtonText("No")
                        .setWarning()
                        .onClick(() => {
                            resolve(false);
                            this.close();
                        })
                    );
                }
            }

            new MyBoolInputModal(app).open();
        });
    }

}

File: insert_card.py
import requests
import json

settings={"language":"Italian"}

url="https://api.magicthegathering.io/v1/cards"

card_name=input("Input card name>")

print("Searching...")

params = {
    'name': card_name,
    "language": settings["language"]
}

response = requests.get(url,params=params)

print("-"*len("Searching..."))

if response.status_code == 200:
    try:
        # Parse the JSON response
        json_data = response.json()

        with open("latest.json","w") as jf:
            json.dump(json_data,jf)

        output_dict={}

        if len(json_data["cards"])==0:
            print("No results :(")
            quit(0)

        for card in json_data["cards"]:
            if settings["language"]!="English" and "foreignNames" not in card:
                continue
            localized_data=[value for value in card["foreignNames"] if value["language"]==settings["language"]]

            if len(localized_data)==0:
                localized_data=card
            else:
                localized_data=localized_data[0]

            output_dict["name"]=localized_data["name"]
            output_dict["rarity"]=card["rarity"]
            output_dict["text"]=localized_data["text"].replace("\"","'").replace("\n"," // ")
            output_dict["types"]=card["types"]
            output_dict["image_url"]=localized_data["imageUrl"]

            print(output_dict["name"])
            print(output_dict["types"])

            if "Land" in card["types"]: # The card is a land
                pass
            elif "Creature" in card["types"]: # The card is a creature
                output_dict["mana_cost"]=card["manaCost"]
                output_dict["power"]=card["power"]
                output_dict["toughness"]=card["toughness"]
                
                print(output_dict["mana_cost"])
                print(f"{output_dict['power']}/{output_dict['toughness']}")
            else:
                output_dict["mana_cost"]=card["manaCost"]
                
                print(output_dict["mana_cost"])

            print(output_dict["text"])

            print("-"*len("Searching..."))
            accepted=input("Accept? [y/n]>")
            print("-"*len("Searching..."))
            if accepted=="y":
                break

        if accepted!="y":
            quit(0)

        if "mana_cost" in output_dict:
            mana_cost=output_dict["mana_cost"]

            if mana_cost[1]not in ["W","B","R","G","U"]: # The first value is a number
                output_dict["neutral"]=int(mana_cost.split("{")[1][:-1])

            output_dict["white"]=mana_cost.count("{W}")
            output_dict["black"]=mana_cost.count("{B}")
            output_dict["red"]=mana_cost.count("{R}")
            output_dict["green"]=mana_cost.count("{G}")
            output_dict["blue"]=mana_cost.count("{U}")

        output_text="---\n"

        for key in output_dict:
            if isinstance(output_dict[key],str):
                output_text+=f"{key}: \"{output_dict[key]}\"\n"
            else:
                output_text+=f"{key}: {output_dict[key]}\n"

        output_text+="---"

        with open(f"Cards/{output_dict["name"]}.md","w") as jf:
            jf.write(output_text)

    except ValueError:
        print("Response content is not valid JSON")
else:
    print(f'Failed to retrieve the data. Status code: {response.status_code}')
